evm/src/main/scala/pravda/evm/debug/evm/EvmDebugger.scala

```diff
@@ -0,0 +1,99 @@
+final case class PravdaOpLog(op: String, snapshot: MemorySnapshot, storage: StorageSnapshot) extends DebugLog
+final case class ErrorLog(op: String, snapshot: MemorySnapshot, storage: StorageSnapshot)    extends DebugLog
+
+final case class MemorySnapshot(stack: List[Data.Primitive], heap: List[Data])
+final case class StorageSnapshot(items: Map[Data.Primitive, Data])
+
+object EvmDebugger extends Debugger[DebugLog] {
+  override def debugOp(program: ByteBuffer, op: Int, mem: MemoryImpl, storage: StorageSandbox)(
+      exec: Either[Throwable, ExecutionResult]): DebugLog = {
+    val memorySnap = MemorySnapshot(mem.stack.toList, mem.heap.toList)
+    val storageSnap = StorageSnapshot(storage.storageItems.toMap)
+    exec match {
+      case Right(UnitExecution(_)) =>
+        PravdaOpLog(mnemonicByOpcode(op), memorySnap, storageSnap)
+      case Right(ex @ MetaExecution(Custom(s))) if s.startsWith(EvmDebugTranslator.debugMarker) =>
+        EvmOpLog(s.stripPrefix(EvmDebugTranslator.debugMarker))
+      case Right(ex @ MetaExecution(l)) =>
+        PravdaOpLog(l.toString, memorySnap, storageSnap)
+      case Right(InterruptedExecution) =>
+        PravdaOpLog(mnemonicByOpcode(op), memorySnap, storageSnap)
+      case Left(e: Data.DataException) =>
+        ErrorLog(s"${mnemonicByOpcode(op)} - ${e.toString}", memorySnap, storageSnap)
+      case Left(ThrowableVmError(e)) =>
+        ErrorLog(s"${mnemonicByOpcode(op)} - ${e.toString}", memorySnap, storageSnap)
+    }
+  }
+
+  private def t(count: Int)(s: String) = "\t" * count + s
+
+  def DebugLogShow(showStack: Boolean, showHeap: Boolean, showStorage: Boolean): cats.Show[DebugLog] = {
+    case EvmOpLog(s) =>
+      s
+    case PravdaOpLog(op, memSnap, storSnap) =>
+      (s"\t\t$op" ::
+        (if (showStack)
+         t(3)(s"stack size = ${memSnap.stack.size}:") :: memSnap.stack.reverse.map(el => t(4)(el.toString))
+       else Nil) :::
+        (if (showHeap) t(3)(s"heap size = ${memSnap.heap.size}:") :: memSnap.heap.map(el => t(4)(el.toString))
+       else Nil)) :::
+        (if (showStorage)
+         t(3)(s"storage size = ${storSnap.items.size}:") :: storSnap.items
+           .map(el => t(4)(s"${el._1} -> ${el._2}"))
+           .toList
+       else Nil) mkString "\n"
+    case ErrorLog(msg, memSnap, storSnap) =>
+      (s"\t\t$msg" ::
+        (if (showStack)
+         t(3)(s"stack size = ${memSnap.stack.size}:") :: memSnap.stack.reverse.map(el => t(4)(el.toString))
+       else Nil) :::
+        (if (showHeap) t(3)(s"heap size = ${memSnap.heap.size}:") :: memSnap.heap.map(el => t(4)(el.toString))
+       else Nil)) :::
+        (if (showStorage)
+         t(3)(s"storage size = ${storSnap.items.size}:") :: storSnap.items
+           .map(el => t(4)(s"${el._1} -> ${el._2}"))
+           .toList
+       else Nil) mkString "\n"
+  }
+
+  def showDebugLogContainer(implicit showDebugLog: Show[DebugLog]): cats.Show[List[DebugLog]] =
+    l =>
```

Comment:
`.map(...).mkString`?

Reply:
It isn't flexible. Better to use type class and configure output in it

Reply:
I mean you can rewrite your code to more readable verison